core/module_graph.h

```diff
@@ -0,0 +1,121 @@
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#ifndef BESS_MODULE_GRAPH_H_
+#define BESS_MODULE_GRAPH_H_
+
+#include <map>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+
+#include "utils/common.h"
+
+class Module;
+class ModuleBuilder;
+
+// Represents a node in `module_graph_`.
+class Node {
+ public:
+  // Creates a new Node that represents `module_`.
+  Node(Module *module) : module_(module), children_() {}
+
+  // Add a child to the node.
+  bool AddChild(const std::string &child) {
+    return children_.insert(child).second;
+  }
+
+  // Remove a child from the node.
+  void RemoveChild(const std::string &child) { children_.erase(child); }
+
+  const Module *module() const { return module_; }
+  const std::unordered_set<std::string> &children() const { return children_; }
+
+ private:
+  // Module that this Node represents.
+  Module *module_;
+
+  // Children of `module_` in the pipeline.
+  std::unordered_set<std::string> children_;
+
+  DISALLOW_COPY_AND_ASSIGN(Node);
+};
+
+class ModuleGraph {
+ public:
+  // Return true if any module from the builder exists
+  static bool ExistModuleClass(const ModuleBuilder *);
+
+  // Add a module to the collection. Returns true on success, false otherwise.
+  static bool AddModule(Module *m);
+
+  // Remove a module from the collection. Returns 0 on success, -errno
+  // otherwise.
+  static int DestroyModule(Module *m, bool erase = true);
+  static void DestroyAllModules();
+
+  static const std::map<std::string, Module *> &GetAllModules();
+
+  static std::string GenerateDefaultName(const std::string &class_name,
+                                         const std::string &default_template);
+
+  // Connects two modules (`to` and `from`) together in `module_graph_`.
+  static bool AddEdge(const std::string &from, const std::string &to);
+
+  // Disconnects two modules (`to` and `from`) together in `module_graph_`.
+  static bool RemoveEdge(const std::string &from, const std::string &to);
+
+ private:
+  // Updates the parents of modules with tasks by traversing `module_graph_` and
+  // ignoring all modules that are not tasks.
+  static bool UpdateTaskGraph();
+
+  // Finds the next module that implements a task, and updates it's parents
```

Comment:
I'm not sure I understand from this comment what this method does.  (I know the comment was already there, just figured we might clarify it.)

Minor: its